internal/errorlint/errorlint.go

```diff
@@ -0,0 +1,191 @@
+	file.walk(func(node ast.Node) bool {
+		call, ok := node.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+
+		if isErrorPackage(call.Fun) || len(call.Args) < 1 {
+			return true
+		}
+
+		msgIndex := 1
+		if isDotInPkg(call.Fun, "errors", "New") || isDotInPkg(call.Fun, "fmt", "Errorf") ||
+			isDotInPkg(call.Fun, "errors", "Errorf") {
+			msgIndex = 0
+		}
+
+		msg, ok := call.Args[msgIndex].(*ast.BasicLit)
+		if !ok || msg.Kind != token.STRING {
+			return true
+		}
+
+		msgString, err := strconv.Unquote(msg.Value)
+		if err != nil {
+			return false
+		}
+
+		if msgString == "" {
+			return true
+		}
+		errormsg := getErrorMessage(msgString)
+		if errormsg != "" {
+			pkgName := getPkgName(call.Fun)
+			pass.Reportf(node.Pos(), "%s "+errormsg, pkgName)
+		}
+		return true
+	})
+}
+
+// isErrorPackage checks if the ast.Expr package matches the error/fmt/trace/log packages for linter
+func isErrorPackage(expr ast.Expr) bool {
+	return !isDotInPkg(expr, "errors", "New") && !isDotInPkg(expr, "errors", "Wrap") &&
+		!isDotInPkg(expr, "errors", "Wrapf") && !isDotInPkg(expr, "log", "Warn") &&
+		!isDotInPkg(expr, "log", "Info") && !isDotInPkg(expr, "log", "Error") &&
+		!isDotInPkg(expr, "trace", "StartSpan") && !isDotInPkg(expr, "trace", "StartCall") &&
+		!isDotInPkg(expr, "fmt", "Errorf") && !isDotInPkg(expr, "errors", "Errorf") &&
+		!isDotInPkg(expr, "errors", "WithMessage") && !isDotInPkg(expr, "errors", "WithMessagef")
+}
+
+// getErrorMessage returns message based on whether it has capitalization, punctuation or not
+func getErrorMessage(msg string) string {
+	isCap, isPunct := isStringFormatted(msg)
+	var errormsg string
+	switch {
+	case isCap && isPunct:
+		errormsg = "message should not be capitalized and should not end with punctuation"
+	case isCap:
+		errormsg = "message should not be capitalized"
+	case isPunct:
+		errormsg = "message should not end with punctuation"
+	default:
+		errormsg = ""
+	}
+
+	return errormsg
+}
+
+// isIdent checks if ident string is equal to ast.Ident name
+func isIdent(expr ast.Expr, ident string) bool {
+	id, ok := expr.(*ast.Ident)
+	return ok && id.Name == ident
+}
+
+// hasDotInPkg checks if pkg.function format is followed
+func isDotInPkg(expr ast.Expr, pkg, name string) bool {
+	sel, ok := expr.(*ast.SelectorExpr)
+	return ok && isIdent(sel.X, pkg) && isIdent(sel.Sel, name)
+}
+
+// getPkgName returns package name errors/fmt/log/trace
+func getPkgName(expr ast.Expr) string {
+	sel, ok := expr.(*ast.SelectorExpr)
+	for _, pkg := range []string{"errors", "log", "fmt", "trace"} {
+		if ok && isIdent(sel.X, pkg) {
+			return pkg
+		}
+	}
+
+	return ""
+}
+
+// isStringFormatted examines error/trace/log strings for incorrect ending and capitalization
+func isStringFormatted(msg string) (isCap, isPunct bool) {
+	last, _ := utf8.DecodeLastRuneInString(msg)
+	isPunct = last == '.' || last == ':' || last == '!' || last == '\n'
+	for _, ch := range msg {
+		if unicode.IsUpper(ch) {
+			isCap = true
+		}
+	}
```

Comment:
shouldn't you just be checking the first character for capitalization?

Reply:
+1 to George's comment. This should only check the first letter.

Reply:
So this is where maybe other people's standards and ORCS standards diverge.  We make the entire error message lowercase.  What is the value in only the first letter being lowercase?

Reply:
the things that come to my mind immediately are acronyms (ASCII, etc) and abbreviations (ID, etc) and initialisms (UUID, etc)

Reply:
FWIW i do see where you're going though if we forget about what i just mentioned above @clevelittlefield-outreach

Reply:
yeah people resist on the acronyms, but the whole goal of this is to always find that error in the code/logs and not worry about casing, the first letter alone doesnt provide value

Reply:
i understand where you're coming from. i'm not the owner of this repo anymore (i just know the most about it) and im not on DT, so ultimately I'm going to defer this one over to @jkinkead

Reply:
I love the idea of having standardized style. I _don't_ love the idea of having _bespoke_ standardized style. For a widely-used linter, we should be adhering to established third-party standards.

Google's style guide (what I suggest we use, and what is linked in the linter) says:

> Error strings should not be capitalized (unless beginning with an exported name, a proper noun or an acronym) and should not end with punctuation. This is because error strings usually appear within other context before being printed to the user.

It further says:

> On the other hand, the style for the full displayed message (logging, test failure, API response, or other UI) depends, but should typically be capitalized.

I don't think this should deviate from this style guide, unless we can find an official Golang one that says something different.

Teams that want a different style that's still compatible with this (e.g. all characters lowercase), that's fine, but I don't think the linter should enforce that rule.

Reply:
So that exception will be hard to codify  (unless beginning with an exported name, a proper noun or an acronym), and I dont think we want to have people do nolints to enforce that exception.  That rule (not uppercase for first character only) honestly seem like arbitrary aesthetics rather than serving a clear purpose.  

I do not understand the further part, we have one string here, how can it be all capitalized in another context?

Beyond that, I swore that gobox (back in go-outreach days) at one time in the readme asked for all lowercase errors. Using that starter guidance, but also building around that idea, is we want all strings of this sort to be able to be found going from code to datadog and from datadog to code.  One of those at the time was case sensitive.  I would have to test again to see if that is still the case.  The point of leaving out casing it so avoid those accidental misses due to case.

If we decide this linter is not right for all of outreach, that is ok. I questioned that when we said it could be outreach wide in the first place because I have seen plenty of code that violates that.  We will also have lots of violations when we build a linter for [this guideline](https://github.com/getoutreach/gobox) as well. Which circles back to the original question, how can we build and include in the ruleset a linter just for just the ORCS team?

Reply:
@jkinkead thoughts?